aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py

```diff
@@ -0,0 +1,828 @@
+            return True
+    return False
+
+
+async def filter_schema(
+    credentials: Sequence[VCRecord], schemas: Sequence[SchemaInputDescriptor]
+) -> Sequence[VCRecord]:
+    """
+    Filter by schema.
+
+    Returns list of credentials where credentialSchema.id or types matched
+    with input_descriptors.schema.uri
+
+    Args:
+        credentials: list of VCRecords to check
+        schemas: list of schemas from the input_descriptors
+    Return:
+        Sequence of filtered VCRecord
+
+    """
+    result = []
+    for credential in credentials:
+        applicable = False
+        for schema in schemas:
+            applicable = await credential_match_schema(
+                credential=credential, schema_id=schema.uri
+            )
+            if schema.required and not applicable:
+                break
+        if applicable:
+            result.append(credential)
+    return result
+
+
+async def credential_match_schema(credential: VCRecord, schema_id: str) -> bool:
+    """
+    Credential matching by schema.
+
+    Used by filter_schema to check if credential.schema_ids or credential.types
+    matched with schema_id
+
+    Args:
+        credential: VCRecord to check
+        schema_id: schema uri to check
+    Return:
+        bool
+    """
+    for cred_schema_id in credential.schema_ids:
+        if cred_schema_id == schema_id:
+            return True
+    for cred_type in credential.types:
+        if cred_type == schema_id:
+            return True
+    return False
+
+
+async def apply_requirements(req: Requirement, credentials: Sequence[VCRecord]) -> dict:
+    """
+    Apply Requirement.
+
+    Args:
+        req: Requirement
+        credentials: Sequence of credentials to check against
+    Return:
+        dict of input_descriptor ID key to list of credential_json
+    """
+    # Dict for storing descriptor_id keys and list of applicable
+    # credentials values
+    result = {}
+    # Get all input_descriptors attached to the PresentationDefinition
+    descriptor_list = req.input_descriptors or []
+    for descriptor in descriptor_list:
+        # Filter credentials to apply filtering upon by matching each credentialSchema.id
+        # or expanded types on each InputDescriptor's schema URIs
+        filtered_by_schema = await filter_schema(
+            credentials=credentials, schemas=descriptor.schemas
+        )
+        # Filter credentials based upon path expressions specified in constraints
+        filtered = await filter_constraints(
+            constraints=descriptor.constraint, credentials=filtered_by_schema
+        )
+        if len(filtered) != 0:
+            result[descriptor._id] = filtered
+
+    if len(descriptor_list) != 0:
+        # Applies min, max or count attributes of submission_requirement
+        if is_len_applicable(req, len(result)):
+            return result
+        return {}
+
+    nested_result = []
+    tmp_dict = {}
```

Comment:
The last tmp_dict. Maybe something like this?
```suggestion
    given_id_descriptors = {}
```